src/metabase/notification/payload/impl/card.clj

```
@@ -11,11 +11,15 @@
 
 (mu/defmethod notification.payload/payload :notification/card
   [{:keys [creator_id payload subscriptions] :as _notification-info} :- ::notification.payload/Notification]
-  (let [card_id (:card_id payload)]
-    {:card_part       (notification.execute/execute-card creator_id card_id
-                                                         ;; for query_execution's context purposes
-                                                         ;; TODO: check whether we can remove this or rename it?
-                                                         :pulse-id (:id payload))
+  (let [card_id     (:card_id payload)
```

The fix is we throw an exception here. why throw an exception?
- well I expected `notification.execute/execute-card` to throw an exception when it executes an invalid card, but since it uses `qp/userland-query` , which run the QP with the [around-middleware](https://github.com/metabase/metabase/blob/better-task-history-report-for-failed-alerts/src/metabase/query_processor.clj#L24), in which it has the `catch-exceptions` middleware. This middleware catch all exceptions and return the error in the result map.
- By throwing an error here, this will buble up the stack and fail at https://github.com/metabase/metabase/blob/23ad35455600c7024c8a5cd636304d2a45ae1ccb/src/metabase/notification/send.clj#L146, which will cause the whole `notification-send` task fail

IMO, this is not the best fix and I'm still open for ideas. But I plan to a refactor of how we handle error across the notification in https://www.notion.so/metabase/Maintenance-alerts-implement-proper-error-handling-and-channel-selection-behavior-for-Send-now--19d69354c901802b944cd0d58940fbd7?pvs=4

---

> IMO, this is not the best fix

Why not though?

---

it's uncoordinated and feel random, IMO we should have something like the error type system from QP. 

notification-payload -> generate the result, if has error, return error as map.

then notification.send check for these result and handle those errors.

---

looks like @crisptrutski has the same idea: https://github.com/metabase/metabase/pull/54949#discussion_r1988761245